src/useReactions.tsx

```diff
@@ -0,0 +1,244 @@
+/*
+Copyright 2024 Milton Moura <miltonmoura@gmail.com>
+
+SPDX-License-Identifier: AGPL-3.0-only
+Please see LICENSE in the repository root for full details.
+*/
+
+import {
+  EventType,
+  MatrixEvent,
+  RelationType,
+  RoomEvent as MatrixRoomEvent,
+} from "matrix-js-sdk/src/matrix";
+import { ReactionEventContent } from "matrix-js-sdk/src/types";
+import {
+  createContext,
+  useContext,
+  useState,
+  ReactNode,
+  useCallback,
+  useEffect,
+  useMemo,
+} from "react";
+import { MatrixRTCSession } from "matrix-js-sdk/src/matrixrtc/MatrixRTCSession";
+import { logger } from "matrix-js-sdk/src/logger";
+
+import { useMatrixRTCSessionMemberships } from "./useMatrixRTCSessionMemberships";
+import { useClientState } from "./ClientContext";
+
+interface ReactionsContextType {
+  raisedHands: Record<string, Date>;
+  addRaisedHand: (userId: string, info: RaisedHandInfo) => void;
+  removeRaisedHand: (userId: string) => void;
+  supportsReactions: boolean;
+  myReactionId: string | null;
+}
+
+const ReactionsContext = createContext<ReactionsContextType | undefined>(
+  undefined,
+);
+
+interface RaisedHandInfo {
+  /**
+   * Call membership event that was reacted to.
+   */
+  membershipEventId: string;
+  /**
+   * Event ID of the reaction itself.
+   */
+  reactionEventId: string;
+  /**
+   * The time when the reaction was raised.
+   */
+  time: Date;
+}
+
+export const useReactions = (): ReactionsContextType => {
+  const context = useContext(ReactionsContext);
+  if (!context) {
+    throw new Error("useReactions must be used within a ReactionsProvider");
+  }
+  return context;
+};
+
+/**
+ * Provider that handles raised hand reactions for a given `rtcSession`.
+ */
+export const ReactionsProvider = ({
+  children,
+  rtcSession,
+}: {
+  children: ReactNode;
+  rtcSession: MatrixRTCSession;
+}): JSX.Element => {
+  const [raisedHands, setRaisedHands] = useState<
+    Record<string, RaisedHandInfo>
+  >({});
+  const memberships = useMatrixRTCSessionMemberships(rtcSession);
+  const clientState = useClientState();
+  const supportsReactions =
+    clientState?.state === "valid" && clientState.supportedFeatures.reactions;
+  const room = rtcSession.room;
+
+  // Calculate our own reaction event.
+  const myReactionId = useMemo((): string | null => {
+    const myUserId = room.client.getUserId();
+    if (myUserId) {
+      return raisedHands[myUserId]?.reactionEventId;
```

Comment:
```suggestion
      return raisedHands[myUserId]?.reactionEventId ?? null;
```

TypeScript lets this break the `string | null` type contract because it thinks `raisedHands[myUserId]` can never be `undefined`